app/services/hashtag_autocomplete_service.rb

```diff
@@ -0,0 +1,187 @@
+            limit: limit,
+            for_input: true,
+          )
+        TagsController
+          .tag_counts_json(tags_with_counts)
+          .take(limit)
+          .map do |tag|
+            HashtagItem.new.tap do |item|
+              item.text = "#{tag[:name]} x #{tag[:count]}"
+              item.slug = tag[:name]
+              item.icon = "tag"
+            end
+          end
+      else
+        []
+      end
+    end
+  end
+
+  clear_data_sources
+
+  class HashtagItem
+    # The text to display in the UI autocomplete menu for the item.
+    attr_accessor :text
+
+    # Canonical slug for the item. Different from the ref, which can
+    # have the type as a suffix to distinguish between conflicts.
+    attr_accessor :slug
+
+    # The icon to display in the UI autocomplete menu for the item.
+    attr_accessor :icon
+
+    # Distinguishes between different entities e.g. tag, category.
+    attr_accessor :type
+
+    # Inserted into the textbox when an autocomplete item is selected,
+    # and must be unique so it can be used for lookups via the #lookup
+    # method above.
+    attr_accessor :ref
+  end
+
+  def initialize(guardian)
+    @guardian = guardian
+  end
+
+  def lookup(slugs)
+    all_slugs = []
+    tag_slugs = []
+
+    slugs[0..HashtagAutocompleteService::HASHTAGS_PER_REQUEST].each do |slug|
+      if slug.end_with?(PrettyText::Helpers::TAG_HASHTAG_POSTFIX)
+        tag_slugs << slug.chomp(PrettyText::Helpers::TAG_HASHTAG_POSTFIX)
+      else
+        all_slugs << slug
+      end
+    end
+
+    # Try to resolve hashtags as categories first
+    category_slugs_and_ids =
+      all_slugs.map { |slug| [slug, Category.query_from_hashtag_slug(slug)&.id] }.to_h
+    category_ids_and_urls =
+      Category
+        .secured(guardian)
+        .select(:id, :slug, :parent_category_id) # fields required for generating category URL
+        .where(id: category_slugs_and_ids.values)
+        .map { |c| [c.id, c.url] }
+        .to_h
+    categories_hashtags = {}
+    category_slugs_and_ids.each do |slug, id|
+      if category_url = category_ids_and_urls[id]
+        categories_hashtags[slug] = category_url
+      end
+    end
+
+    # Resolve remaining hashtags as tags
+    tag_hashtags = {}
+    if SiteSetting.tagging_enabled
+      tag_slugs += (all_slugs - categories_hashtags.keys)
+      DiscourseTagging
+        .filter_visible(Tag.where_name(tag_slugs), guardian)
+        .each { |tag| tag_hashtags[tag.name] = tag.full_url }
+    end
+
+    { categories: categories_hashtags, tags: tag_hashtags }
+  end
+
+  def search(term, types_in_priority_order, limit = 5)
+    results = []
+    slugs_by_type = {}
+    term = term.downcase
+    types_in_priority_order =
+      types_in_priority_order.select { |type| @@data_sources.keys.include?(type) }
+
+    types_in_priority_order.each do |type|
+      data = @@data_sources[type].call(guardian, term, limit - results.length)
+      if !data.all? { |item|
```

Comment:
would probably use do...end form here

Reply:
I think syntax tree actually did this haha, will manually change but it's something we should look out for. I've just been using it occasionally for brand new files in core